ledger/src/shred.rs

```diff
@@ -524,6 +495,73 @@ impl Shred {
     }
 }
 
+// Helper methods to extract pieces of the shred from the payload
+// without deserializing the entire payload.
+pub mod layout {
+    use {super::*, std::ops::Range};
+
+    fn get_shred_size(packet: &Packet) -> usize {
+        if packet.meta.repair() {
+            packet.meta.size.saturating_sub(SIZE_OF_NONCE)
+        } else {
+            packet.meta.size
+        }
+    }
+
+    pub fn get_shred(packet: &Packet) -> &[u8] {
+        &packet.data[..get_shred_size(packet)]
+    }
+
+    pub(crate) fn get_signature(shred: &[u8]) -> Option<Signature> {
+        Some(Signature::new(shred.get(..SIZE_OF_SIGNATURE)?))
+    }
+
+    pub(crate) const fn get_signature_range() -> Range<usize> {
+        0..SIZE_OF_SIGNATURE
+    }
+
+    pub(super) fn get_shred_type(shred: &[u8]) -> Result<ShredType, Error> {
+        match shred.get(OFFSET_OF_SHRED_TYPE) {
+            None => Err(Error::InvalidPayloadSize(shred.len())),
+            Some(shred_type) => match ShredType::try_from(*shred_type) {
+                Err(_) => Err(Error::InvalidShredType),
+                Ok(shred_type) => Ok(shred_type),
+            },
+        }
+    }
+
+    pub fn get_slot(shred: &[u8]) -> Option<Slot> {
+        deserialize_from_with_limit(shred.get(OFFSET_OF_SHRED_SLOT..)?).ok()
+    }
+
+    pub(super) fn get_index(shred: &[u8]) -> Option<u32> {
+        deserialize_from_with_limit(shred.get(OFFSET_OF_SHRED_INDEX..)?).ok()
+    }
+
+    // Returns chunk of the payload which is signed.
+    pub(crate) fn get_signed_message(shred: &[u8]) -> Option<&[u8]> {
```

Comment:
This may need to return `Option<&[[u8]]>` to account for a non-contiguous buffer if the signed data refers to everything protected by merkle vs just the signature for the merkle root.

Reply:
> if the signed data refers to everything protected by merkle

not sure why should we want to sign everything,
but even then, in https://github.com/solana-labs/solana/pull/25237 that is still a contiguous slice.

Reply:
ok, you can disregard. I was considering support for a shred layout where the merkle hashed data is non-contiguous.